src/iop/retouch.c

```diff
@@ -0,0 +1,4569 @@
+    if (i == p->curr_scale)
+      cairo_set_source_rgb(cr, 1, 1, 0);
+    else if (i >= p->merge_from_scale && i <= p->num_scales && p->merge_from_scale > 0)
+      cairo_set_source_rgb(cr, z * .95, z * .95, 0);
+    else
+      cairo_set_source_rgb(cr, z, z, z);
+    cairo_fill(cr);
+    
+    // if the scale has shapes inform it
+    if (rt_scale_has_shapes(p, i))
+    {
+      cairo_set_source_rgb(cr, .1, .8, 0);
+      cairo_rectangle(cr, (1. / (RETOUCH_NO_SCALES)) * i, 0, (1. / (RETOUCH_NO_SCALES)), 0.1);
+      cairo_fill(cr);
+    }
+  }
+  
+  cairo_set_antialias(cr, CAIRO_ANTIALIAS_DEFAULT);
+  cairo_restore(cr);
+
+  // if mouse is over a box, draw a rectangle
+  if (g->mouse_y > arrw * .5f && g->mouse_y < height - arrw * .5f)
+  {
+    const int curr_scale = rt_mouse_x_to_wdbar_box(g->mouse_x, width);
+    if (curr_scale >= 0 && curr_scale < RETOUCH_NO_SCALES)
+    {
+      cairo_set_line_width(cr, DT_PIXEL_APPLY_DPI(1.));
+      cairo_set_source_rgb(cr, 0.8, 0, 0);
+      cairo_rectangle(cr,
+          (width / (float)RETOUCH_NO_SCALES) * curr_scale + inset + DT_PIXEL_APPLY_DPI(3),
+          inset + arrw * .5f + DT_PIXEL_APPLY_DPI(1),
+          (width / (float)RETOUCH_NO_SCALES) - 2 * DT_PIXEL_APPLY_DPI(3),
+          height - arrw - 2 * DT_PIXEL_APPLY_DPI(1));
+      cairo_stroke(cr);
+    }
+  }
+
+  /* render control points handles */
+  cairo_set_source_rgb(cr, 0.6, 0.6, 0.6);
+  cairo_set_line_width(cr, DT_PIXEL_APPLY_DPI(1.));
+  
+  // draw number of scales arrow (bottom arrow)
+  {
+    const float middle = rt_get_middle_wdbar_box((float)p->num_scales, width);
+    gboolean is_under_mouse = rt_mouse_over_arrow_wdbar((float)p->num_scales, g->mouse_x, width);
+    is_under_mouse &= rt_mouse_over_bottom_wdbar(g->mouse_y, height);
+    
+    cairo_move_to(cr, inset + middle, height + (2 * inset) - 1);
+    cairo_rel_line_to(cr, -arrw * .5f, 0);
+    cairo_rel_line_to(cr, arrw * .5f, -arrw);
+    cairo_rel_line_to(cr, arrw * .5f, arrw);
+    cairo_close_path(cr);
+
+    if(is_under_mouse || g->is_dragging == dt_iop_retouch_drag_bottom)
+      cairo_fill(cr);
+    else
+      cairo_stroke(cr);
+  }
+
+  // draw merge scales arrow (top arrow)
+  {
+    const float middle = rt_get_middle_wdbar_box((float)p->merge_from_scale, width);
+    gboolean is_under_mouse = rt_mouse_over_arrow_wdbar((float)p->merge_from_scale, g->mouse_x, width);
+    is_under_mouse &= rt_mouse_over_top_wdbar(g->mouse_y);
+
+    cairo_move_to(cr, inset + middle, 1);
+    cairo_rel_line_to(cr, -arrw * .5f, 0);
+    cairo_rel_line_to(cr, arrw * .5f, arrw);
+    cairo_rel_line_to(cr, arrw * .5f, -arrw);
+    cairo_close_path(cr);
+
+    if(is_under_mouse || g->is_dragging == dt_iop_retouch_drag_top)
+      cairo_fill(cr);
+    else
+      cairo_stroke(cr);
+  }
+
+  /* push mem surface into widget */
+  cairo_destroy(cr);
+  cairo_set_source_surface(crf, cst, 0, 0);
+  cairo_paint(crf);
+  cairo_surface_destroy(cst);
+
+  return TRUE;
+}
+#endif
+
+#if 0
```

Comment:
I would remove the code or add a comment about why it will be used later. This is large code and it will certainly ease the integration if retouch.c get smaller.